include/fiction/algorithms/simulation/sidb/assess_physical_population_stability.hpp

```diff
@@ -0,0 +1,261 @@
+                    valid_state_counter += 1;
+                    break;
+                }
+            }
+        }
+
+        return valid_states_and_detailed_popstability_information;
+    };
+
+    /**
+     * This function collects the system energy with the corresponding charge index information of all physically valid
+     * charge distributions of a given SiDB layout.
+     *
+     * @param sim_results The simulation results, including all physically valid charge distributions.
+     * @return A vector of pairs, each containing a double value representing the system energy and a uint64_t
+     * representing the unique charge index. The vector is sorted in ascending order of the energy values.
+     */
+    [[nodiscard]] std::vector<std::pair<double, uint64_t>>
+    collect_energy_and_charge_index(const sidb_simulation_result<Lyt>& sim_results) noexcept
+    {
+        std::unordered_map<std::size_t, std::tuple<typename Lyt::cell, transition_type, double>> results{};
+
+        std::vector<std::pair<double, uint64_t>> energy_and_unique_charge_index{};
+
+        std::transform(sim_results.charge_distributions.cbegin(), sim_results.charge_distributions.cend(),
+                       std::back_inserter(energy_and_unique_charge_index),
+                       [](const auto& ch_lyt) {
+                           return std::make_pair(ch_lyt.get_system_energy(), ch_lyt.get_charge_index_and_base().first);
+                       });
+
+        // Sort the vector in ascending order of the double value
+        std::sort(energy_and_unique_charge_index.begin(), energy_and_unique_charge_index.end(),
+                  [](const auto& lhs, const auto& rhs) { return lhs.first < rhs.first; });
+
+        return energy_and_unique_charge_index;
+    }
+
+  private:
+    /**
+     * Layout to analyze.
+     */
+    const Lyt& layout;
+    /**
+     * Parameters used for the physical simulation.
+     */
+    const sidb_simulation_parameters& params;
+};
+
+}  // namespace detail
+
+/**
+ * This function assesses the population stability of each physically valid charge distributions of a given SiDB layout.
+ * It determines the minimum absolute electrostatic potential required to induce a charge distribution transition.
+ * The function also identifies the critical SiDB and the charge state transition (i.e., the change from one charge
+ * state to another).
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ * @param lyt The layout for which the population stability is assessed.
+ * @param params The simulation parameters used for the assessment.
+ * @return An unordered map where each key represents the charge distribution in ascending energy order, and the
+ * value is a structure containing information about the critical SiDB, the type of charge state transition, and the
+ * minimum electrostatic potential required for the transition.
+ */
+template <typename Lyt>
+[[nodiscard]] std::unordered_map<std::size_t, population_stability_information<Lyt>>
```

Comment:
Please make the entire return type into a custom `struct` or `class`. The start with `population_stability_information` is great! If you can now wrap said type into a more intuitive one than "map of numbers", that would be even better! :)